### PR TITLE
Leaderboard

### PR DESCRIPTION
**Pull request by x-tabdeveloping**
This PR includes a preiminary version of the leaderboard that is refitted for the MTEB library's current state and complexity.
I added code for loading and filtering results based on the user's queries interactively, and an elementary table to display results.
To run the leaderboard, you can import it from MTEB:
```python
from mteb.leaderboard import demo

demo.launch()
``` 
Model filtering is not yet implemented due to the `ModelMeta` object's current inadequacy (#1234 ), but I included some code and UI for how I think filtering models should probably work.
The leaderboard is currently written in Gradio and there are a number of challenges that need to be addressed with the UI, most notably scrollbars in the dropdowns, which might contain hundreds of options.
@KennethEnevoldsen and I agreed that we should merge these changes after a review and then subsequently submit issues the leaderboard still has so that multiple people can work on different things simultaneously.

### CODE REVIEW

File: mteb/load_results/benchmark_results.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+from mteb.languages import ISO_LANGUAGE
+from mteb.load_results.task_results import TaskResult
+from mteb.models.overview import get_model_metas
+from mteb.overview import get_tasks
+
+Split = str
+Score = Any
+
+
+def restrict_task_results(res: TaskResult, task: AbsTask) -> TaskResult:
```
Comment (KennethEnevoldsen) [inline mteb/load_results/benchmark_results.py:22]: Is this a copy of existing utility? Did copy it or was it moved?

edit seems like you refactored it

Comment (KennethEnevoldsen) [reply mteb/load_results/benchmark_results.py:22]:
```suggestion
def _restrict_task_results(task_result: TaskResult, task: AbsTask) -> TaskResult:
```

File: mteb/load_results/benchmark_results.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+            )
+        seen_splits.add(split)
+    if seen_splits != set(splits):
+        raise ValueError(f"Missing splits {set(splits) - seen_splits}")
+    new_res = {**res.to_dict(), "scores": new_scores}
+    new_res = TaskResult.from_dict(new_res)
+    return new_res
+
+
+class ModelResult(BaseModel):
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply mteb/load_results/benchmark_results.py:49]: This class looks really good

File: scripts/leaderboard/results_all.csv
Comment (KennethEnevoldsen) [inline scripts/leaderboard/results_all.csv:1]: delete?

File: scripts/leaderboard/app.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  - [ ] Model metadata (embedding size)
+  - [ ] total Co2 emissions
+- [ ] Add results loading from Hub
+- [ ] Benchmark selection
+- [ ] task type selection
+- [ ] domain selection
+- [ ] inidivual task selection
+
+- Optimization to be added:
+  - mteb.load_results is called for each custom language selection. A solution it so load the results once and filter them in the app.
```
Comment (KennethEnevoldsen) [inline scripts/leaderboard/app.py:16]: <3

**Review by KennethEnevoldsen — APPROVED** [review]
This looks really good. Only a few minor comments.

Seems like you need to run `make install`, `make lint` for the linting - also seems like the tests fails due to `Dict` being uppercase (we have dropped support for Python 3.8 due to it reaching end-of-life)

Seems like refactoring the model_meta is gonna be needed. What other large elements are still missing? Or can we start handing out smaller segments?

Seems like we also need some dependencies (I would just specify it as `pip install mteb[leaderboard]`)

File: mteb/leaderboard/utils.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    domains = [
+        task.metadata.domains
+        for task in mteb.get_tasks()
+        if task.metadata.domains is not None
+    ]
+    domains = itertools.chain.from_iterable(domains)
+    return list(set(domains))
+
+
+def equal_criteria(c0, c1):
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply mteb/leaderboard/utils.py:28]: I'm pretty sure I'm not using this anymore, it's from a previous demo iteration

File: mteb/leaderboard/utils.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    keys = set(c0.keys()) | set(c1.keys())
+    for key in keys:
+        if set(c0.get(key, [])) != set(c1.get(key, [])):
+            return False
+    return True
+
+
+def get_model_size_range() -> tuple[int, int]:
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply mteb/leaderboard/utils.py:36]: Yea, can totally be done! I need data on model sizes though. I don't think we have this in ModelMeta yet

File: mteb/load_results/benchmark_results.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply mteb/load_results/benchmark_results.py:49]: Thanks :D Calling it ModelResult was a conscious choice. I don't have strong preferences here though, so we can call it ModelResults if we want to.

File: mteb/leaderboard/table.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import numpy as np
+import pandas as pd
+
+from mteb.overview import get_task
+
+
+def scores_to_table(scores_long: list[dict]):
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply mteb/leaderboard/table.py:7]: I'm not sure. I wouldn't. The `get_scores()` methods currently return something that is really easy to turn into a generic table, but I'm pretty sure the leaderboard is specific enough to be turned into its own thing

File: mteb/leaderboard/utils.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    langs = [task.languages for task in mteb.get_tasks()]
+    langs = itertools.chain.from_iterable(langs)
+    return list(set(langs))
+
+
+def get_task_types():
+    return list(set(task.metadata.type for task in mteb.get_tasks()))
+
+
+def get_domains():
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply mteb/leaderboard/utils.py:18]: I don't think I'm using this anywhere. Will delete

File: mteb/leaderboard/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply mteb/leaderboard/utils.py:28]: I will delete it

**Comment by x-tabdeveloping** [comment]
Some of the bigger pieces we might be missing from MTEB (the library itself):
1. I'm not sure `restrict_task_results` works the way it should. As far as I've seen the `scores` property still includes other languages, and when I call `get_score()` on a `TaskResult` object with a different set of languages, I get a slightly different result. This might just be that we don't remove things from the `languages` field, but it's definitely worth investigating. I will dig a bit more.
2. ModelMeta needs serious work.
3. We need to speed things up. Filtering and serializing/deserializing result objects takes way longer than it should and it will seriously hinder users' experience. I know this is inconvenient but we might have to find faster ways of validating objects than Pydantic in some places. ([msgpack](https://github.com/jcrist/msgspec) might help us out here.)

**Comment by x-tabdeveloping** [comment]
Work the leaderboard needs:
1. We need to figure something out for the scrollable dropdown. I tried custom CSS (`overflow-y: scroll`) but scrollbars seems to have been tempered with by the Gradio folks, and while it limits the size of the dropdown, the scroll doesn't work at all.
2. We need a better and smarter table (links, etc.) perhaps even pagination. The table layout also needs work, as the table keeps shrinking if you scroll up and down. I have no idea what is causing this.
3. We need callbacks for filtering models, this should be easy once we have ModelMetas ready.

File: scripts/leaderboard/app.py (hunk 3 above)
Comment (x-tabdeveloping) [reply scripts/leaderboard/app.py:16]: <3

**Comment by x-tabdeveloping** [comment]
I'm at a bit of a loss as to what's causing the test to fail, seems quite unrelated to me

**Comment by x-tabdeveloping** [comment]
Yea, here's a reproducible example of the problem I've been talking about:

```python
import mteb
from mteb.load_results.benchmark_results import _restrict_task_results

results = mteb.load_results()

task_res = results[0][10]
print(task_res.languages)
# ['deu', 'ita', 'eng', 'fra']

task = mteb.get_task("RTE3", languages=["deu"])
print(task.languages)
# ['deu']

restricted = _restrict_task_results(task_res, task)
print(restricted.languages)
# ['deu', 'ita', 'eng', 'fra']

print(restricted.get_score() == task_res.get_score())
# True

print(restricted.get_score(languages="deu") == restricted.get_score())
# False

```

**Comment by x-tabdeveloping** [comment]
Okay I think I fixed it, but beware that this issue might still affect `get_task()` and `get_tasks()`.

**Comment by x-tabdeveloping** [comment]
hmmm actually, I didn't

**Comment by x-tabdeveloping** [comment]
Okay, now it works properly.
However, tests are still not passing and is due to the fact that `CrossEncoder` objects in `sentence-transformers` do not have a `model_card_data` attribute, not even in version 3.0.0.
I can do a hotfix, but we'll have to do something about this.

**Review by KennethEnevoldsen — APPROVED** [review]

**Comment by isaac-chung** [comment]
> Okay, now it works properly.
> However, tests are still not passing and is due to the fact that `CrossEncoder` objects in `sentence-transformers` do not have a `model_card_data` attribute, not even in version 3.0.0.
> I can do a hotfix, but we'll have to do something about this.

@x-tabdeveloping here is the related issue: https://github.com/embeddings-benchmark/mteb/issues/1192

**Pull request merged**